lib/array.rb

```diff
@@ -0,0 +1,52 @@
+class Array
+
+
+  def my_inject(sum=first,method=nil)
+
+    if method == nil && sum != first
+      if !block_given?
+      method = sum
+      sum = first
+      end
+    end
+    if method == nil
+      raise 'no block given' if !block_given? && method == nil
+      each do |x|
+        if index(x) == 0 && sum != first
+          sum = yield sum,x
+        end
+        if index(x) != 0
+          sum = yield sum,x
+        else
+          sum
+        end
+      end
+    else
+      each do |x|
+        if index(x) == 0 && sum != first
+          sum = x.send(method,sum)
+        end
+        if index(x) != 0
+          sum = x.send(method,sum)
+        end
+      end
+    end
+    sum
+  end
+
```

Comment:
Be wary of nested `if` statements - this might mean delegating to private methods, as you did in your Takeaway, or a rethinking of the entire structure of the function